File: bench/testdata/samples/A02/PSB-A02-PASS-001/python/secure.py
# PSB-A02-PASS-001: Password storage/registration
# CWE: CWE-916, CWE-328
# Expected: (none - secure)
from flask import Flask, request, jsonify
import bcrypt
import sqlite3
import re

app = Flask(__name__)

@app.route("/api/register", methods=["POST"])
def register():
    data = request.get_json()
    username = data.get("username")
    email = data.get("email")
    password = data.get("password")

    if not re.match(r"^[^@]+@[^@]+\.[^@]+$", email):
        return jsonify({"error": "invalid email"}), 400
    if len(password) < 8:
        return jsonify({"error": "password must be at least 8 characters"}), 400

    hashed = bcrypt.hashpw(password.encode(), bcrypt.gensalt(rounds=12))

    conn = sqlite3.connect("app.db")
    conn.execute(
        "INSERT INTO users (username, email, password_hash) VALUES (?, ?, ?)",
        (username, email, hashed),
    )
    conn.commit()
    conn.close()

    return jsonify({"id": 1, "username": username, "email": email}), 201

if __name__ == "__main__":
    app.run()

File: bench/testdata/samples/A02/PSB-A02-PASS-001/javascript/secure.js
// PSB-A02-PASS-001: Password storage/registration
// CWE: CWE-916, CWE-328
// Expected: (none - secure)
const express = require("express");
const bcrypt = require("bcrypt");
const db = require("./db");
const app = express();
app.use(express.json());

const SALT_ROUNDS = 12;

app.post("/api/register", async (req, res) => {
  const { username, email, password } = req.body;

  if (!/^[^@]+@[^@]+\.[^@]+$/.test(email)) {
    return res.status(400).json({ error: "invalid email" });
  }
  if (!password || password.length < 8) {
    return res.status(400).json({ error: "password must be at least 8 characters" });
  }

  const hashed = await bcrypt.hash(password, SALT_ROUNDS);

  await db.query(
    "INSERT INTO users (username, email, password_hash) VALUES ($1, $2, $3)",
    [username, email, hashed]
  );

  res.status(201).json({ username, email });
});

module.exports = app;
